kube-controllers/pkg/controllers/loadbalancer/loadbalancer_controller.go

```diff
@@ -625,69 +580,116 @@ func (c *loadBalancerController) resolvePools(pools []string, isv4 bool) ([]cnet
+	loadBalancerIPs := []cnet.IP{}
+	ipv4Pools := []cnet.IPNet{}
+	ipv6Pools := []cnet.IPNet{}
+
+	for key, annotation := range annotations {
+		switch key {
+		case annotationIpv4Pools:
+			poolCIDRs := []string{}
+			err := json.Unmarshal([]byte(annotation), &poolCIDRs)
+			if err != nil {
+				return nil, nil, nil, err
+			}
+			ipv4Pools, err = c.resolvePools(poolCIDRs, true)
+			if err != nil {
+				return nil, nil, nil, err
+			}
+			if ipv4Pools == nil {
+				return nil, nil, nil, fmt.Errorf("failed to resolve pools for IPv4 addresses from annotation")
+			}
+		case annotationIpv6Pools:
+			poolCIDRs := []string{}
+			err := json.Unmarshal([]byte(annotation), &poolCIDRs)
+			if err != nil {
+				return nil, nil, nil, err
+			}
+			ipv6Pools, err = c.resolvePools(poolCIDRs, false)
+			if err != nil {
+				return nil, nil, nil, err
+			}
+			if ipv6Pools == nil {
+				return nil, nil, nil, fmt.Errorf("failed to resolve pools for IPv6 addresses from annotation")
+			}
+		case annotationLoadBalancerIp:
+			ipAddrs := []string{}
+			err := json.Unmarshal([]byte(annotation), &ipAddrs)
+			if err != nil {
+				return nil, nil, nil, err
+			}
+			ipv4 := 0
+			ipv6 := 0
+			for _, ipAddr := range ipAddrs {
+				curr := cnet.ParseIP(ipAddr)
+				if curr == nil {
+					return nil, nil, nil, errors.New(fmt.Sprintf("Could not parse %s as a valid IP address", ipAddr))
+				}
+				if curr.To4() != nil {
+					ipv4++
+				} else if curr.To16() != nil {
+					ipv6++
+				}
+				loadBalancerIPs = append(loadBalancerIPs, *curr)
+			}
 
-	if ipv6 > 1 || ipv4 > 1 {
-		return nil, errors.New(fmt.Sprintf("At max only one ipv4 and one ipv6 address can be specified. Recieved %d ipv4 and %d ipv6 addresses", ipv4, ipv6))
+			if ipv6 > 1 || ipv4 > 1 {
+				return nil, nil, nil, errors.New(fmt.Sprintf("At max only one ipv4 and one ipv6 address can be specified. Recieved %d ipv4 and %d ipv6 addresses", ipv4, ipv6))
+			}
+		}
 	}
-	return parsedIPs, nil
-}
-
-func parseAnnotation(str string) []string {
-	str = strings.TrimPrefix(str, "[")
-	str = strings.TrimSuffix(str, "]")
-	return strings.Split(str, ",")
+	return loadBalancerIPs, ipv4Pools, ipv6Pools, nil
 }
 
 // createHandle Returns a handle to use for IP allocation for the service
-func createHandle(svc *v1.Service) string {
-	handle := strings.ToLower(fmt.Sprintf("lb.%s.%s", svc.Name, svc.UID))
-	if len(handle) > k8svalidation.DNS1123SubdomainMaxLength {
-		return handle[:k8svalidation.DNS1123SubdomainMaxLength]
+func createHandle(svc *v1.Service) (string, error) {
+	prefix := "lb-"
+	handle := strings.ToLower(fmt.Sprintf("%s-%s-%s", svc.Name, svc.Namespace, svc.UID))
+
+	hasher := sha256.New()
+	_, err := hasher.Write([]byte(handle))
+	if err != nil {
+		log.WithError(err).Panic("Failed to generate hash from handle")
+		return "", err
+	}
+
+	hash := base64.RawURLEncoding.EncodeToString(hasher.Sum(nil))
+	regex := regexp.MustCompile("([-_.])")
+	hash = regex.ReplaceAllString(hash, "")
+	handle = prefix + hash
```

Comment:
This is fine, and no need to change I don't think - but what I was suggesting was that we only add the hash if the handle exceeds the name limit.

e.g.,

```go
handle := strings.ToLower(fmt.Sprintf("%s-%s-%s", svc.Name, svc.Namespace, svc.UID))

if len(handle) > k8svalidation.DNS1123LabelMaxLength {
  // Handle is too long - trim off the end, leaving enough room for 8 bytes of hash.
  hash := generateHash(handle)
  handle = handle[:k8svalidation.DNS1123LabelMaxLength-8]
  handle = fmt.Sprintf("%s%s", handle, hash[:8])
}
```

This  just makes the actual handle IDs human readable in most cases.